ponix4-src/source/xorg-ponix4/driver-git-12122012/xf86-input-keyboard/man/kbd.4
.\" shorthand for double quote that works everywhere.
.ds q \N'34'
.TH KBD 4 "xf86-input-keyboard 1.6.2" "X Version 11"
.SH NAME
kbd \- Keyboard input driver
.SH SYNOPSIS
.nf
.B "Section \*qInputDevice\*q"
.BI "  Identifier \*q" idevname \*q
.B  "  Driver \*qkbd\*q"
\ \ ...
.B EndSection
.fi
.SH DESCRIPTION
.B kbd 
is an Xorg input driver for keyboards.  The driver supports the standard
OS-provided keyboard interface, but these are currently only available to
this driver module for Linux, BSD, and Solaris.  This driver is the
replacement for the built-in
.B keyboard
driver formerly included in Xorg.
.PP
The
.B kbd
driver functions as a keyboard input device.
.SH CONFIGURATION DETAILS
.PP
Depending on the X server version in use, input device options may be set
in either a xorg.conf file, an xorg.conf.d snippet, or in the
configuration files read by the Hardware Abstraction Layer (HAL) daemon,
hald(1).
.PP
Please refer to xorg.conf(5) for general configuration
details and for options that can be used with all input drivers.  This
section only covers configuration details specific to this driver.
.PP
The following driver
.B Options
are supported:
.TP 7
.BI "Option \*qDevice\*q \*q" string \*q
Specify the keyboard device.  Default: the OS's default console keyboard
input source.  Property: "Device Node" (read-only).
.TP 7
.BI "Option \*qProtocol\*q \*q" string \*q
Specify the keyboard protocol.  Valid protocol types include:
.PP
.RS 12
Standard, Xqueue.
.RE
.PP
.RS 7
Not all protocols are supported on all platforms.  Default: "Standard".
.RE
.TP 7
.BI "Option \*qXLeds\*q \*q" ledlist \*q
makes the keyboard LEDs specified in
.I ledlist
available for client use instead of their traditional function
(Scroll Lock, Caps Lock and Num Lock).  The numbers in the list are
in the range 1 to 3.  Default: empty list.
.TP 7
.BI "Option \*qXkbRules\*q \*q" rules \*q
specifies which XKB rules file to use for interpreting the
.BR XkbModel ,
.BR XkbLayout ,
.BR XkbVariant ,
and
.B XkbOptions
settings.  Default: "base" for most platforms.
If you use the "base" value then you can
find listing of all valid values for these four options in the
.I
/usr/xorg/share/X11/xkb/rules/base.lst
file.
.TP 7
.BI "Option \*qXkbModel\*q \*q" modelname \*q
specifies the XKB keyboard model name.  Default: "pc105" for most platforms.
.TP 7
.BI "Option \*qXkbLayout\*q \*q" layoutname \*q
specifies the XKB keyboard layout name.  This is usually the country or
language type of the keyboard.  Default: "us" for most platforms.
.TP 7
.BI "Option \*qXkbVariant\*q \*q" variants \*q
specifies the XKB keyboard variant components.  These can be used to
enhance the keyboard layout details.  Default: not set.
.TP 7
.BI "Option \*qXkbOptions\*q \*q" options \*q
specifies the XKB keyboard option components.  These can be used to
enhance the keyboard behaviour.  Default: not set.
.PP
For a list of available XKB options, see
.B xkeyboard-config(7).
.SH EXAMPLE
The following xorg.conf fragment ensures that user will be able to switch between
.I us
and
.I sk
layouts by pressing the "menu" key. The scroll lock LED shows which layout
is currently active.
.PP
The
.B XkbVariant
option defines which variants of the two layouts should be used. In case of the
.I us
layout
its default variant is used. In case of the
.I sk
layout
its
.I qwerty
variant is used.
.PP
.nf
.RS 4
Section \*qInputDevice\*q
.RS 4
Identifier   "Generic Keyboard"
Driver       "kbd"
Option       "CoreKeyboard"
Option       "XkbRules"      "base"
Option       "XkbModel"      "pc105"
Option       "XkbLayout"     "us,sk"
Option       "XkbVariant"    ",qwerty"
Option       "XkbOptions"    "grp:menu_toggle,grp_led:scroll"
.RE
EndSection
.RE
.fi
.SH "SEE ALSO"
Xorg(1), xorg.conf(5), 
Xserver(1), X(7).

hal(7), hald(8), fdi(5).

xkeyboard-config(7).


ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/xorg-ponix4/driver-git-12122012/xf86-input-keyboard/src/kbd.c
/*
 * Copyright (c) 2002 by The XFree86 Project, Inc.
 * Author: Ivan Pascal.
 *
 * Based on the code from
 * xf86Config.c which is
 * Copyright 1991-2002 by The XFree86 Project, Inc.
 * Copyright 1997 by Metro Link, Inc.
 * xf86Events.c and xf86Io.c which are
 * Copyright 1990,91 by Thomas Roell, Dinkelscherben, Germany.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <xorg-server.h>

#include <X11/X.h>
#include <X11/Xproto.h>

#include "xf86.h"
#include "atKeynames.h"
#include "xf86Privstr.h"

#include <X11/extensions/XI.h>
#include <X11/extensions/XIproto.h>
#include "extnsionst.h"
#include "extinit.h"
#include "inputstr.h"

#include "xf86Xinput.h"
#include "xf86_OSproc.h"
#include "xf86OSKbd.h"
#include "compiler.h"

#include "exevents.h"
#include <X11/Xatom.h>
#include "xserver-properties.h"

#include "xkbstr.h"
#include "xkbsrv.h"

#define CAPSFLAG	1
#define NUMFLAG		2
#define SCROLLFLAG	4
#define MODEFLAG	8
#define COMPOSEFLAG	16
/* Used to know when the first DEVICE_ON after a DEVICE_INIT is called */
#define INITFLAG	(1U << 31)

#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) < 12
static InputInfoPtr KbdPreInit(InputDriverPtr drv, IDevPtr dev, int flags);
#else
static int KbdPreInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags);
#endif
static int KbdProc(DeviceIntPtr device, int what);
static void KbdCtrl(DeviceIntPtr device, KeybdCtrl *ctrl);
static void KbdBell(int percent, DeviceIntPtr dev, pointer ctrl, int unused);
static void PostKbdEvent(InputInfoPtr pInfo, unsigned int key, Bool down);

static void InitKBD(InputInfoPtr pInfo, Bool init);
static void UpdateLeds(InputInfoPtr pInfo);

_X_EXPORT InputDriverRec KBD = {
	1,
	"kbd",
	NULL,
	KbdPreInit,
	NULL,
	NULL
};

_X_EXPORT InputDriverRec KEYBOARD = {
	1,
	"keyboard",
	NULL,
	KbdPreInit,
	NULL,
	NULL
};

static const char *kbdDefaults[] = {
    "Protocol",		"standard",
    "XkbRules",		"base",
    "XkbModel",		"pc105",
    "XkbLayout",	"us",
    NULL
};

static char *xkb_rules;
static char *xkb_model;
static char *xkb_layout;
static char *xkb_variant;
static char *xkb_options;

#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) < 12
static int
NewKbdPreInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags);

static InputInfoPtr
KbdPreInit(InputDriverPtr drv, IDevPtr dev, int flags)
{
    InputInfoPtr pInfo;

    if (!(pInfo = xf86AllocateInput(drv, 0)))
	return NULL;

    pInfo->name = dev->identifier;
    pInfo->flags = XI86_KEYBOARD_CAPABLE;
    pInfo->conversion_proc = NULL;
    pInfo->reverse_conversion_proc = NULL;
    pInfo->private_flags = 0;
    pInfo->always_core_feedback = NULL;
    pInfo->conf_idev = dev;
    pInfo->close_proc = NULL;

    if (NewKbdPreInit(drv, pInfo, flags) == Success)
    {
        pInfo->flags |= XI86_CONFIGURED;
        return pInfo;
    }

    xf86DeleteInput(pInfo, 0);
    return NULL;
}

static int
NewKbdPreInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
#else
static int
KbdPreInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
#endif
{
    KbdDevPtr pKbd;
    char *s;
    const char **defaults;
    int rc = Success;

    /* Initialise the InputInfoRec. */
    pInfo->type_name = XI_KEYBOARD;
    pInfo->device_control = KbdProc;
    /*
     * We don't specify our own read_input function. We expect
     * an OS specific readInput() function to handle this.
     */
    pInfo->read_input = NULL;
    pInfo->control_proc = NULL;
    pInfo->switch_mode = NULL;
    pInfo->fd = -1;
    pInfo->dev = NULL;

    defaults = kbdDefaults;
    xf86CollectInputOptions(pInfo, defaults
#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) < 12
            , NULL
#endif
            );
    xf86ProcessCommonOptions(pInfo, pInfo->options); 

    if (!(pKbd = calloc(sizeof(KbdDevRec), 1))) {
        rc = BadAlloc;
        goto out;
    }

    pInfo->private = pKbd;
    pKbd->PostEvent = PostKbdEvent;

    if (!xf86OSKbdPreInit(pInfo)) {
        rc = BadAlloc;
        goto out;
    }

    if (!pKbd->OpenKeyboard(pInfo)) {
        rc = BadMatch;
        goto out;
    }

    if ((s = xf86SetStrOption(pInfo->options, "XLeds", NULL))) {
        char *l, *end;
        unsigned int i;
        l = strtok(s, " \t\n");
        while (l) {
    	    i = strtoul(l, &end, 0);
    	    if (*end == '\0')
    	        pKbd->xledsMask |= 1L << (i - 1);
    	    else {
    	        xf86Msg(X_ERROR, "\"%s\" is not a valid XLeds value", l);
    	    }
    	    l = strtok(NULL, " \t\n");
        }
        free(s);
    }

    xkb_rules = xf86SetStrOption(pInfo->options, "XkbRules", NULL);
    xkb_model = xf86SetStrOption(pInfo->options, "XkbModel", NULL);
    xkb_layout = xf86SetStrOption(pInfo->options, "XkbLayout", NULL);
    xkb_variant = xf86SetStrOption(pInfo->options, "XkbVariant", NULL);
    xkb_options = xf86SetStrOption(pInfo->options, "XkbOptions", NULL);

    pKbd->CustomKeycodes = xf86SetBoolOption(pInfo->options, "CustomKeycodes",
                                             FALSE);

out:
  return rc;
}

static void
KbdBell(int percent, DeviceIntPtr dev, pointer ctrl, int unused)
{
   InputInfoPtr pInfo = (InputInfoPtr) dev->public.devicePrivate;
   KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;
   pKbd->Bell(pInfo, percent, ((KeybdCtrl*) ctrl)->bell_pitch,
                              ((KeybdCtrl*) ctrl)->bell_duration);
}

static void
UpdateLeds(InputInfoPtr pInfo)
{
    KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;
    unsigned long leds = 0;

    if (pKbd->keyLeds & CAPSFLAG)    leds |= XLED1;
    if (pKbd->keyLeds & NUMFLAG)     leds |= XLED2;
    if (pKbd->keyLeds & SCROLLFLAG ||
        pKbd->keyLeds & MODEFLAG)    leds |= XLED3;
    if (pKbd->keyLeds & COMPOSEFLAG) leds |= XLED4;

    pKbd->leds = (pKbd->leds & pKbd->xledsMask) | (leds & ~pKbd->xledsMask);
    pKbd->SetLeds(pInfo, pKbd->leds);
}

static void
KbdCtrl( DeviceIntPtr device, KeybdCtrl *ctrl)
{
   unsigned long leds;
   InputInfoPtr pInfo = (InputInfoPtr) device->public.devicePrivate;
   KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;

   if ( ctrl->leds & XLED1) {
       pKbd->keyLeds |= CAPSFLAG;
   } else {
       pKbd->keyLeds &= ~CAPSFLAG;
   }
   if ( ctrl->leds & XLED2) {
       pKbd->keyLeds |= NUMFLAG;
   } else {
       pKbd->keyLeds &= ~NUMFLAG;
   }
   if ( ctrl->leds & XLED3) {
       pKbd->keyLeds |= SCROLLFLAG;
   } else {
       pKbd->keyLeds &= ~SCROLLFLAG;
   }
   if ( ctrl->leds & (XCOMP|XLED4) ) {
       pKbd->keyLeds |= COMPOSEFLAG;
   } else {
       pKbd->keyLeds &= ~COMPOSEFLAG;
   }
   leds = ctrl->leds & ~(XCAPS | XNUM | XSCR); /* ??? */
   pKbd->leds = leds;
  pKbd->SetLeds(pInfo, pKbd->leds);
}

static void
InitKBD(InputInfoPtr pInfo, Bool init)
{
  KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;

  pKbd->scanPrefix      = 0;

  if (init) {
      pKbd->keyLeds = pKbd->GetLeds(pInfo);
      UpdateLeds(pInfo);
      pKbd->keyLeds |= INITFLAG;
  } else {
      unsigned long leds = pKbd->keyLeds;

      pKbd->keyLeds = pKbd->GetLeds(pInfo);
      UpdateLeds(pInfo);
      if ((pKbd->keyLeds & CAPSFLAG) !=
	  ((leds & INITFLAG) ? 0 : (leds & CAPSFLAG))) {
	  pKbd->PostEvent(pInfo, KEY_CapsLock, TRUE);
	  pKbd->PostEvent(pInfo, KEY_CapsLock, FALSE);
      }
      if ((pKbd->keyLeds & NUMFLAG) !=
	  (leds & INITFLAG ? 0 : leds & NUMFLAG)) {
	  pKbd->PostEvent(pInfo, KEY_NumLock, TRUE);
	  pKbd->PostEvent(pInfo, KEY_NumLock, FALSE);
      }
  }
}

static int
KbdProc(DeviceIntPtr device, int what)
{

  InputInfoPtr pInfo = device->public.devicePrivate;
  KbdDevPtr pKbd = (KbdDevPtr) pInfo->private;
  KeySymsRec           keySyms;
  CARD8                modMap[MAP_LENGTH];
  int                  ret;

  switch (what) {
     case DEVICE_INIT:
         ret = pKbd->KbdInit(pInfo, what);
         if (ret != Success)
             return ret;

         pKbd->KbdGetMapping(pInfo, &keySyms, modMap);

         device->public.on = FALSE;
#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) >= 5
         {
             XkbRMLVOSet rmlvo;
             rmlvo.rules = xkb_rules;
             rmlvo.model = xkb_model;
             rmlvo.layout = xkb_layout;
             rmlvo.variant = xkb_variant;
             rmlvo.options = xkb_options;

             if (!InitKeyboardDeviceStruct(device, &rmlvo, KbdBell, KbdCtrl))
             {
                 xf86Msg(X_ERROR, "%s: Keyboard initialization failed. This "
                         "could be a missing or incorrect setup of "
                         "xkeyboard-config.\n", device->name);

                 return BadValue;
             }
         }
# ifdef XI_PROP_DEVICE_NODE
         {
             const char *device_node =
                 xf86CheckStrOption(pInfo->options, "Device", NULL);

             if (device_node)
             {
                 Atom prop_device = MakeAtom(XI_PROP_DEVICE_NODE,
                                             strlen(XI_PROP_DEVICE_NODE), TRUE);
                 XIChangeDeviceProperty(device, prop_device, XA_STRING, 8,
                                        PropModeReplace, strlen(device_node),
                                        device_node, FALSE);
             }
         }
# endif /* XI_PROP_DEVICE_NODE */
#else
         {
             XkbComponentNamesRec xkbnames;
             memset(&xkbnames, 0, sizeof(xkbnames));
             XkbSetRulesDflts(xkb_rules, xkb_model, xkb_layout,
                              xkb_variant, xkb_options);
             XkbInitKeyboardDeviceStruct(device, &xkbnames, &keySyms,
                                         modMap, KbdBell,
                                         (KbdCtrlProcPtr)KbdCtrl);
         }
#endif /* XINPUT ABI 5*/
         InitKBD(pInfo, TRUE);
         break;
  case DEVICE_ON:
    if (device->public.on)
	break;
    /*
     * Set the keyboard into "direct" mode and turn on
     * event translation.
     */
    if ((ret = pKbd->KbdOn(pInfo, what)) != Success)
	return ret;
    /*
     * Discard any pending input after a VT switch to prevent the server
     * passing on parts of the VT switch sequence.
     */
    if (pInfo->fd >= 0) {
	xf86FlushInput(pInfo->fd);
	AddEnabledDevice(pInfo->fd);
    }

    device->public.on = TRUE;
    InitKBD(pInfo, FALSE);
    break;

  case DEVICE_CLOSE:
  case DEVICE_OFF:

    /*
     * Restore original keyboard directness and translation.
     */
    if (pInfo->fd != -1)
      RemoveEnabledDevice(pInfo->fd);
    pKbd->KbdOff(pInfo, what);
    device->public.on = FALSE;
    break;

  default:
    return BadValue;
  }
  return (Success);
}

static void
PostKbdEvent(InputInfoPtr pInfo, unsigned int scanCode, Bool down)
{

  KbdDevPtr    pKbd = (KbdDevPtr) pInfo->private;
  DeviceIntPtr device = pInfo->dev;
  KeyClassRec  *keyc = device->key;
  int state;

#ifdef DEBUG
  ErrorF("kbd driver rec scancode: 0x02%x %s\n", scanCode, down?"down":"up");
#endif
	  
  /*
   * First do some special scancode remapping ...
   */
  if (pKbd->RemapScanCode != NULL) {
     if (pKbd->RemapScanCode(pInfo, (int*) &scanCode))
         return;
  } else {
     if (pKbd->scancodeMap != NULL) {
         TransMapPtr map = pKbd->scancodeMap; 
         if (scanCode >= map->begin && scanCode < map->end)
             scanCode = map->map[scanCode - map->begin];
     }
  }

  /*
   * PC keyboards generate separate key codes for
   * Alt+Print and Control+Pause but in the X keyboard model
   * they need to get the same key code as the base key on the same
   * physical keyboard key.
   */

  state = XkbStateFieldFromRec(&keyc->xkbInfo->state);

  if (((state & AltMask) == AltMask) && (scanCode == KEY_SysReqest))
    scanCode = KEY_Print;
  else if (scanCode == KEY_Break)
    scanCode = KEY_Pause;

  xf86PostKeyboardEvent(device, scanCode + MIN_KEYCODE, down);
}

static void
xf86KbdUnplug(pointer	p)
{
}

static pointer
xf86KbdPlug(pointer	module,
	    pointer	options,
	    int		*errmaj,
	    int		*errmin)
{
    xf86AddInputDriver(&KBD, module, 0);

    return module;
}

static XF86ModuleVersionInfo xf86KbdVersionRec =
{
    "kbd",
    MODULEVENDORSTRING,
    MODINFOSTRING1,
    MODINFOSTRING2,
    XORG_VERSION_CURRENT,
    PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR, PACKAGE_VERSION_PATCHLEVEL,
    ABI_CLASS_XINPUT,
    ABI_XINPUT_VERSION,
    MOD_CLASS_XINPUT,
    {0, 0, 0, 0}		/* signature, to be patched into the file by */
				/* a tool */
};

_X_EXPORT XF86ModuleData kbdModuleData = {
    &xf86KbdVersionRec,
    xf86KbdPlug,
    xf86KbdUnplug
};
